# Description
Detect a person's sentiment in text using Unsupervised Sentiment Neuron, a model trained on 82 million Amazon reviews. It's able to predict sentiment by taking in a string, and outputting the sentiment value from -1 to 1. The full text of the review is used to detect sentiment, so even if the sentiment switches mid-text, it'll still be detected.

**Possible Use Cases**
1. Process tweets or comments to see what positive or negative things people are saying about your product.
2. Quickly filter survey results to pick the most positive or negative feedback
3. Process transcripts to understand how someone is feeling throughout a conversation.

# Examples

Input Text | Sentiment Output (From -1 to 1) | Good or Bad (As an Example) |
--- |   --- | --- |
the food was hot and the service was fast! | 0.5305157 | 👍 |
worst thing ever, i hate everything about this, will return immediately | -1.6204989 | 👎 |

### Keep 👀 reading 👀 to see how to get these results!

# How Good is This Model?

This model has been trained on 82 million Amazon reviews, taking 1 month across 4 GPUs. As with any ML model, the underlying training corpus will affect how effective this model will be in your specific use case, but the model is relatively generalizable outside of reviews. It achieves near state of the art performance, with only 7.7% error, whereas the [state of the art](https://arxiv.org/abs/1605.07725) achieves 5.91% accuracy on the IMDB sentiment dataset.

## Load Code & Weights From Github Repo

In [0]:
!git clone https://github.com/openai/generating-reviews-discovering-sentiment.git

## Change Directory to Cloned Repo

In [0]:
%cd generating-reviews-discovering-sentiment

## Install Dependencies

In [0]:
!pip install tqdm

## Initialize the Model

The model will automatically load all the weights from the `/generating-reviews-discovering-sentiment/model/` directory.

In [0]:
from encoder import Model

model = Model()

## Load some text examples and infer the sentiment

In [17]:
text = [
    'the food was hot and the service was fast!',
    'worst thing ever, i hate everything about this, will return immediately',
    'dead on arrival',
    'this is a life saver, i cant live without this',
    'it was okay, nothing amazing, but nothing to complain about either'
]
text_features = model.transform(text)

18.492 seconds to transform 5 examples


## Print The Sentiment Values

In [21]:
# Why 2388? See: https://github.com/openai/generating-reviews-discovering-sentiment/issues/2
print(text_features[:, 2388])

[ 0.5305157  -1.6204989  -0.6985707   1.7084963  -0.03653394]


**Visualizing Output in Table Form, For Your Convenience**

Input Text | Sentiment (From -1 to 1) |
--- |   ---: |
the food was hot and the service was fast! | 0.5305157 |
worst thing ever, i hate everything about this, will return immediately | -1.6204989 |
dead on arrival | -0.6985707 |
this is a life saver, i cant live without this | 1.7084963 |
it was okay, nothing amazing, but nothing to complain about either | -0.03653394 |

## References
- [Learning to Generate Reviews and Discovering Sentiment](https://arxiv.org/abs/1704.01444) (Alec Radford, Rafal Jozefowicz, Ilya Sutskever)
- [Unsupervised Sentiment Neuron Blog Post](https://blog.openai.com/unsupervised-sentiment-neuron/)